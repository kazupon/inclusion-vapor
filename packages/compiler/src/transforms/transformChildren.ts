// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `@vue/compiler-vapor`
// Author: Evan you (https://github.com/yyx990803) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/vuejs/core-vapor
// Code url: https://github.com/vuejs/core-vapor/blob/6608bb31973d35973428cae4fbd62026db068365/packages/compiler-vapor/src/transforms/transformChildren.ts

import { IRNodeTypes, DynamicFlag } from '../ir'
import { transformNode } from '../transform'

import type { NodeTransform } from './types'

export const transformChildren: NodeTransform = (node, context) => {
  const isFragment =
    node.type === IRNodeTypes.ROOT ||
    node.type === 'Fragment' ||
    node.type === 'Element' ||
    node.type === 'InlineComponent'
  if (!isFragment) {
    return
  }

  const children = node.children || []
  for (const [i, child] of children.entries()) {
    const childContext = context.create(child, i)
    transformNode(childContext)

    if (isFragment) {
      childContext.reference()
      childContext.registerTemplate()

      if (
        !(childContext.dynamic.flags & DynamicFlag.NON_TEMPLATE) ||
        childContext.dynamic.flags & DynamicFlag.INSERT
      ) {
        context.block.returns.push(childContext.dynamic.id!)
      }
    } else {
      context.childrenTemplate.push(childContext.template)
    }

    context.dynamic.children[i] = childContext.dynamic
  }
}
