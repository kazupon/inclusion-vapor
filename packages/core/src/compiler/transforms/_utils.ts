import { parse } from 'svelte/compiler'
import { generate } from '@vue-vapor/compiler-vapor'
import { transform } from '../transform'
import { IRNodeTypes } from '../ir'

import type {
  RootIRNode as VaporRootIRNode,
  CompilerOptions as VaporCompilerOptions
} from '@vue-vapor/compiler-vapor'
import type { CompilerOptions } from '../compile'
import type { RootNode } from '../ir'

export const DEFAULT_OPTIONS: CompilerOptions = {
  prefixIdentifiers: true
}
export const DEFAULT_VAPOR_COMPILER_OPTIONS = DEFAULT_OPTIONS as VaporCompilerOptions

export function makeCompile(options: CompilerOptions = {}) {
  return (source: string, overrideOptions: CompilerOptions = {}) => {
    const svelteAst = parse(source)

    const ast: RootNode = {
      type: IRNodeTypes.ROOT,
      children: svelteAst.html.children || [],
      source,
      components: [],
      directives: [],
      helpers: new Set(),
      temps: 0
    }

    const ir = transform(ast, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })
    const { code, helpers, vaporHelpers } = generate(ir as unknown as VaporRootIRNode, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })

    return { ast, ir, code, helpers, vaporHelpers }
  }
}
