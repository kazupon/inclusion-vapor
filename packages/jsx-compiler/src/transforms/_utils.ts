import { parse } from '@babel/parser'
import { generate } from '@vue-vapor/compiler-vapor'
import { transform } from '../transform'
import { IRNodeTypes } from '../ir'

import type {
  RootIRNode as VaporRootIRNode,
  CompilerOptions as VaporCompilerOptions
} from '@vue-vapor/compiler-vapor'
import type { CompilerOptions } from '../compile'
import type { RootNode, JSXElement, JSXFragment } from '../ir'

export const DEFAULT_OPTIONS: CompilerOptions = {
  prefixIdentifiers: true
}
export const DEFAULT_VAPOR_COMPILER_OPTIONS = DEFAULT_OPTIONS as VaporCompilerOptions

export function makeCompile(options: CompilerOptions = {}) {
  return (source: string, overrideOptions: CompilerOptions = {}) => {
    const {
      body: [statement]
    } = parse(source, {
      sourceType: 'module',
      plugins: ['jsx']
    }).program
    let children!: JSXElement[] | JSXFragment['children']
    if (statement.type === 'ExpressionStatement') {
      children =
        statement.expression.type === 'JSXFragment'
          ? statement.expression.children
          : (statement.expression.type === 'JSXElement'
            ? [statement.expression]
            : [])
    }

    const ast: RootNode = {
      type: IRNodeTypes.ROOT,
      children,
      source,
      components: [],
      directives: [],
      helpers: new Set(),
      temps: 0
    }

    const ir = transform(ast, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })
    const { code, helpers, vaporHelpers } = generate(ir as unknown as VaporRootIRNode, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })

    return { ast, ir, code, helpers, vaporHelpers }
  }
}
