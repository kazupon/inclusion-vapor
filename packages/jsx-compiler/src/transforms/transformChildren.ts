// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `unplugin-vue-jsx-vapor`
// Author: zhiyuanzmj (https://github.com/zhiyuanzmj) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/unplugin/unplugin-vue-jsx-vapor
// Code url: https://github.com/unplugin/unplugin-vue-jsx-vapor/tree/main/src/core/compiler

import { isComponentNode } from './utils'
import { IRNodeTypes, DynamicFlag } from '../ir'
import { transformNode } from '../transform'

import type { BabelNode, IRDynamicInfo } from '../ir'
import type { NodeTransform } from './types'
import type { TransformContext } from './context'

export const transformChildren: NodeTransform = (node, context) => {
  const isFragment =
    node.type === IRNodeTypes.ROOT || node.type === 'JSXFragment' || isComponentNode(node)

  if (node.type !== 'JSXElement' && !isFragment) {
    return
  }

  ;[...node.children].forEach((child, index) => {
    if (child.type === 'JSXText' && !child.value.trim()) {
      child.value = ' '
      if (!index) {
        node.children.splice(0, 1)
      } else if (index === node.children.length) {
        node.children.splice(-1, 1)
      }
    }
  })

  for (const [i, child] of node.children.entries()) {
    const childContext = context.create(child, i)
    transformNode(childContext)

    if (isFragment) {
      childContext.reference()
      childContext.registerTemplate()

      if (
        !(childContext.dynamic.flags & DynamicFlag.NON_TEMPLATE) ||
        childContext.dynamic.flags & DynamicFlag.INSERT
      ) {
        context.block.returns.push(childContext.dynamic.id!)
      }
    } else {
      context.childrenTemplate.push(childContext.template)
    }

    context.dynamic.children[i] = childContext.dynamic
  }

  if (!isFragment) {
    processDynamicChildren(context as TransformContext<BabelNode>)
  }
}

function processDynamicChildren(context: TransformContext<BabelNode>) {
  let prevDynamics: IRDynamicInfo[] = []
  let hasStaticTemplate = false
  const children = context.dynamic.children

  for (const [index, child] of children.entries()) {
    if (child.flags & DynamicFlag.INSERT) {
      prevDynamics.push(child)
    }

    if (!(child.flags & DynamicFlag.NON_TEMPLATE)) {
      if (prevDynamics.length > 0) {
        if (hasStaticTemplate) {
          context.childrenTemplate[index - prevDynamics.length] = `<!>`

          prevDynamics[0].flags -= DynamicFlag.NON_TEMPLATE
          const anchor = (prevDynamics[0].anchor = context.increaseId())

          context.registerOperation({
            type: IRNodeTypes.INSERT_NODE,
            elements: prevDynamics.map(child => child.id!),
            parent: context.reference(),
            anchor
          })
        } else {
          context.registerOperation({
            type: IRNodeTypes.PREPEND_NODE,
            elements: prevDynamics.map(child => child.id!),
            parent: context.reference()
          })
        }
        prevDynamics = []
      }
      hasStaticTemplate = true
    }
  }

  if (prevDynamics.length > 0) {
    context.registerOperation({
      type: IRNodeTypes.INSERT_NODE,
      elements: prevDynamics.map(child => child.id!),
      parent: context.reference()
    })
  }
}
