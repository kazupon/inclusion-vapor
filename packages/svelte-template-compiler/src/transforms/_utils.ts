import { parse } from 'svelte/compiler'
import { generate } from '@vue-vapor/compiler-vapor'
import { transform } from '../transform'
import { IRNodeTypes } from '../ir'

import type {
  RootIRNode as VaporRootIRNode,
  CompilerOptions as VaporCompilerOptions
} from '@vue-vapor/compiler-vapor'
import type { CompilerOptions } from '../compile'
import type { RootNode, RootIRNode } from '../ir'

export const DEFAULT_OPTIONS: CompilerOptions = {
  prefixIdentifiers: true
}
export const DEFAULT_VAPOR_COMPILER_OPTIONS = DEFAULT_OPTIONS as VaporCompilerOptions

export function makeCompile(options: CompilerOptions = {}) {
  return (
    source: string,
    overrideOptions: CompilerOptions = {}
  ): {
    ast: RootNode
    ir: RootIRNode
    code: string
    helpers: Set<string>
    vaporHelpers: Set<string>
  } => {
    const svelteAst = parse(source)

    const ast: RootNode = {
      type: IRNodeTypes.ROOT,
      children: svelteAst.html.children || [],
      source,
      components: [],
      directives: [],
      helpers: new Set(),
      temps: 0
    }

    const ir = transform(ast, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })
    const { code, helpers, vaporHelpers } = generate(ir as unknown as VaporRootIRNode, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })

    return { ast, ir, code, helpers, vaporHelpers }
  }
}
