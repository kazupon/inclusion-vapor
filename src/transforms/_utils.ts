import { parse } from 'svelte/compiler'
import { generate } from '@vue-vapor/compiler-vapor'
import { transform } from '../transform'
import { IRNodeTypes } from '../ir'

import type { RootIRNode as VaporRootIRNode } from '@vue-vapor/compiler-vapor'
import type { CompilerOptions } from '../compile'
import type { RootNode } from '../ir'

export function makeCompile(options: CompilerOptions = {}) {
  return (source: string, overrideOptions: CompilerOptions = {}) => {
    const svelteAst = parse(source)

    const ast: RootNode = {
      type: IRNodeTypes.ROOT,
      children: svelteAst.html.children || [],
      source,
      components: [],
      directives: [],
      helpers: new Set(),
      temps: 0
    }

    const ir = transform(ast, {
      prefixIdentifiers: true,
      ...options,
      ...overrideOptions
    })
    const { code, helpers, vaporHelpers } = generate(ir as unknown as VaporRootIRNode, {
      prefixIdentifiers: true,
      ...options,
      ...overrideOptions
    })

    return { ast, ir, code, helpers, vaporHelpers }
  }
}
