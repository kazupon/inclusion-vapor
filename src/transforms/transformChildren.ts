import { IRNodeTypes, DynamicFlag } from '../ir'
import { transformNode } from '../transform'

import type { NodeTransform } from './types'

export const transformChildren: NodeTransform = (node, context) => {
  const isFragment =
    node.type === IRNodeTypes.ROOT || node.type === 'Fragment' || node.type === 'InlineComponent'
  if (!isFragment) {
    return
  }

  const children = node.children || []
  for (const [i, child] of children.entries()) {
    const childContext = context.create(child, i)
    transformNode(childContext)

    if (isFragment) {
      childContext.reference()
      childContext.registerTemplate()

      if (
        !(childContext.dynamic.flags & DynamicFlag.NON_TEMPLATE) ||
        childContext.dynamic.flags & DynamicFlag.INSERT
      ) {
        context.block.returns.push(childContext.dynamic.id!)
      }
    } else {
      context.childrenTemplate.push(childContext.template)
    }

    context.dynamic.children[i] = childContext.dynamic
  }
}
