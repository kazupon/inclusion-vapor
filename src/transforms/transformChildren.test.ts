import { describe, expect, test } from 'vitest'
import { makeCompile } from './_utils'
import { transformChildren } from './transformChildren'

const compileWithTransform = makeCompile({
  nodeTransforms: [transformChildren]
})

describe('compiler: children transform', () => {
  test.todo('basic')

  test('native elements', () => {
    const {
      ir,
      code,
      vaporHelpers: _
    } = compileWithTransform(
      `<div>
        <div>
          <p>Hello</p>
        </div>
        <p>World</p>
      </div>`
    )
    expect(code).toMatchSnapshot()
    expect(ir).toMatchSnapshot()
  })

  test('sibling references', () => {
    const {
      ir,
      code,
      vaporHelpers: _
    } = compileWithTransform(
      `<div>
        <p>{ first }</p>
        { second }
        { third }
        <p>{ forth }</p>
      </div>`
    )
    expect(code).toMatchSnapshot()
    expect(ir).toMatchSnapshot()
    // TODO: vaporHelpers
    // expect(Array.from(vaporHelpers)).containSubset([
    //   'next',
    //   'setText',
    //   'createTextNode',
    //   'insert',
    //   'template',
    // ])
  })

  test('components', () => {
    const {
      ir,
      code,
      vaporHelpers: _
    } = compileWithTransform(
      `<div>
        <Foo>{ foo }</Foo>
        <Bar><Baz /></Bar>
      </div>`
    )
    expect(code).toMatchSnapshot()
    expect(ir).toMatchSnapshot()
    // TODO: vaporHelpers
  })
})
