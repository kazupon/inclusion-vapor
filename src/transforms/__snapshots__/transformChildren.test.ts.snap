// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compiler: children transform > components 1`] = `
"
export function render(_ctx) {
  return n0
}"
`;

exports[`compiler: children transform > components 2`] = `
{
  "block": {
    "dynamic": {
      "children": [
        {
          "children": [],
          "flags": 1,
          "id": 0,
        },
      ],
      "flags": 1,
    },
    "effect": [],
    "node": {
      "children": [
        {
          "attributes": [],
          "children": [
            {
              "data": "
        ",
              "end": 14,
              "raw": "
        ",
              "start": 5,
              "type": "Text",
            },
            {
              "attributes": [],
              "children": [
                {
                  "end": 26,
                  "expression": Node {
                    "end": 24,
                    "loc": SourceLocation {
                      "end": Position {
                        "column": 18,
                        "line": 2,
                      },
                      "start": Position {
                        "column": 15,
                        "line": 2,
                      },
                    },
                    "name": "foo",
                    "start": 21,
                    "type": "Identifier",
                  },
                  "start": 19,
                  "type": "MustacheTag",
                },
              ],
              "end": 32,
              "name": "Foo",
              "start": 14,
              "type": "InlineComponent",
            },
            {
              "data": "
        ",
              "end": 41,
              "raw": "
        ",
              "start": 32,
              "type": "Text",
            },
            {
              "attributes": [],
              "children": [
                {
                  "attributes": [],
                  "children": [],
                  "end": 53,
                  "name": "Baz",
                  "start": 46,
                  "type": "InlineComponent",
                },
              ],
              "end": 59,
              "name": "Bar",
              "start": 41,
              "type": "InlineComponent",
            },
            {
              "data": "
      ",
              "end": 66,
              "raw": "
      ",
              "start": 59,
              "type": "Text",
            },
          ],
          "end": 72,
          "name": "div",
          "start": 0,
          "type": "Element",
        },
      ],
      "components": [],
      "directives": [],
      "helpers": Set {},
      "source": "<div>
        <Foo>{ foo }</Foo>
        <Bar><Baz /></Bar>
      </div>",
      "temps": 0,
      "type": 0,
    },
    "operation": [],
    "returns": [
      0,
    ],
    "type": 1,
  },
  "component": Set {},
  "directive": Set {},
  "node": {
    "children": [
      {
        "attributes": [],
        "children": [
          {
            "data": "
        ",
            "end": 14,
            "raw": "
        ",
            "start": 5,
            "type": "Text",
          },
          {
            "attributes": [],
            "children": [
              {
                "end": 26,
                "expression": Node {
                  "end": 24,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 18,
                      "line": 2,
                    },
                    "start": Position {
                      "column": 15,
                      "line": 2,
                    },
                  },
                  "name": "foo",
                  "start": 21,
                  "type": "Identifier",
                },
                "start": 19,
                "type": "MustacheTag",
              },
            ],
            "end": 32,
            "name": "Foo",
            "start": 14,
            "type": "InlineComponent",
          },
          {
            "data": "
        ",
            "end": 41,
            "raw": "
        ",
            "start": 32,
            "type": "Text",
          },
          {
            "attributes": [],
            "children": [
              {
                "attributes": [],
                "children": [],
                "end": 53,
                "name": "Baz",
                "start": 46,
                "type": "InlineComponent",
              },
            ],
            "end": 59,
            "name": "Bar",
            "start": 41,
            "type": "InlineComponent",
          },
          {
            "data": "
      ",
            "end": 66,
            "raw": "
      ",
            "start": 59,
            "type": "Text",
          },
        ],
        "end": 72,
        "name": "div",
        "start": 0,
        "type": "Element",
      },
    ],
    "components": [],
    "directives": [],
    "helpers": Set {},
    "source": "<div>
        <Foo>{ foo }</Foo>
        <Bar><Baz /></Bar>
      </div>",
    "temps": 0,
    "type": 0,
  },
  "source": "<div>
        <Foo>{ foo }</Foo>
        <Bar><Baz /></Bar>
      </div>",
  "template": [],
  "type": 0,
}
`;

exports[`compiler: children transform > sibling references 1`] = `
"
export function render(_ctx) {
  return n0
}"
`;

exports[`compiler: children transform > sibling references 2`] = `
{
  "block": {
    "dynamic": {
      "children": [
        {
          "children": [],
          "flags": 1,
          "id": 0,
        },
      ],
      "flags": 1,
    },
    "effect": [],
    "node": {
      "children": [
        {
          "attributes": [],
          "children": [
            {
              "data": "
        ",
              "end": 14,
              "raw": "
        ",
              "start": 5,
              "type": "Text",
            },
            {
              "attributes": [],
              "children": [
                {
                  "end": 26,
                  "expression": Node {
                    "end": 24,
                    "loc": SourceLocation {
                      "end": Position {
                        "column": 18,
                        "line": 2,
                      },
                      "start": Position {
                        "column": 13,
                        "line": 2,
                      },
                    },
                    "name": "first",
                    "start": 19,
                    "type": "Identifier",
                  },
                  "start": 17,
                  "type": "MustacheTag",
                },
              ],
              "end": 30,
              "name": "p",
              "start": 14,
              "type": "Element",
            },
            {
              "data": "
        ",
              "end": 39,
              "raw": "
        ",
              "start": 30,
              "type": "Text",
            },
            {
              "end": 49,
              "expression": Node {
                "end": 47,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 16,
                    "line": 3,
                  },
                  "start": Position {
                    "column": 10,
                    "line": 3,
                  },
                },
                "name": "second",
                "start": 41,
                "type": "Identifier",
              },
              "start": 39,
              "type": "MustacheTag",
            },
            {
              "data": "
        ",
              "end": 58,
              "raw": "
        ",
              "start": 49,
              "type": "Text",
            },
            {
              "end": 67,
              "expression": Node {
                "end": 65,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 15,
                    "line": 4,
                  },
                  "start": Position {
                    "column": 10,
                    "line": 4,
                  },
                },
                "name": "third",
                "start": 60,
                "type": "Identifier",
              },
              "start": 58,
              "type": "MustacheTag",
            },
            {
              "data": "
        ",
              "end": 76,
              "raw": "
        ",
              "start": 67,
              "type": "Text",
            },
            {
              "attributes": [],
              "children": [
                {
                  "end": 88,
                  "expression": Node {
                    "end": 86,
                    "loc": SourceLocation {
                      "end": Position {
                        "column": 18,
                        "line": 5,
                      },
                      "start": Position {
                        "column": 13,
                        "line": 5,
                      },
                    },
                    "name": "forth",
                    "start": 81,
                    "type": "Identifier",
                  },
                  "start": 79,
                  "type": "MustacheTag",
                },
              ],
              "end": 92,
              "name": "p",
              "start": 76,
              "type": "Element",
            },
            {
              "data": "
      ",
              "end": 99,
              "raw": "
      ",
              "start": 92,
              "type": "Text",
            },
          ],
          "end": 105,
          "name": "div",
          "start": 0,
          "type": "Element",
        },
      ],
      "components": [],
      "directives": [],
      "helpers": Set {},
      "source": "<div>
        <p>{ first }</p>
        { second }
        { third }
        <p>{ forth }</p>
      </div>",
      "temps": 0,
      "type": 0,
    },
    "operation": [],
    "returns": [
      0,
    ],
    "type": 1,
  },
  "component": Set {},
  "directive": Set {},
  "node": {
    "children": [
      {
        "attributes": [],
        "children": [
          {
            "data": "
        ",
            "end": 14,
            "raw": "
        ",
            "start": 5,
            "type": "Text",
          },
          {
            "attributes": [],
            "children": [
              {
                "end": 26,
                "expression": Node {
                  "end": 24,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 18,
                      "line": 2,
                    },
                    "start": Position {
                      "column": 13,
                      "line": 2,
                    },
                  },
                  "name": "first",
                  "start": 19,
                  "type": "Identifier",
                },
                "start": 17,
                "type": "MustacheTag",
              },
            ],
            "end": 30,
            "name": "p",
            "start": 14,
            "type": "Element",
          },
          {
            "data": "
        ",
            "end": 39,
            "raw": "
        ",
            "start": 30,
            "type": "Text",
          },
          {
            "end": 49,
            "expression": Node {
              "end": 47,
              "loc": SourceLocation {
                "end": Position {
                  "column": 16,
                  "line": 3,
                },
                "start": Position {
                  "column": 10,
                  "line": 3,
                },
              },
              "name": "second",
              "start": 41,
              "type": "Identifier",
            },
            "start": 39,
            "type": "MustacheTag",
          },
          {
            "data": "
        ",
            "end": 58,
            "raw": "
        ",
            "start": 49,
            "type": "Text",
          },
          {
            "end": 67,
            "expression": Node {
              "end": 65,
              "loc": SourceLocation {
                "end": Position {
                  "column": 15,
                  "line": 4,
                },
                "start": Position {
                  "column": 10,
                  "line": 4,
                },
              },
              "name": "third",
              "start": 60,
              "type": "Identifier",
            },
            "start": 58,
            "type": "MustacheTag",
          },
          {
            "data": "
        ",
            "end": 76,
            "raw": "
        ",
            "start": 67,
            "type": "Text",
          },
          {
            "attributes": [],
            "children": [
              {
                "end": 88,
                "expression": Node {
                  "end": 86,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 18,
                      "line": 5,
                    },
                    "start": Position {
                      "column": 13,
                      "line": 5,
                    },
                  },
                  "name": "forth",
                  "start": 81,
                  "type": "Identifier",
                },
                "start": 79,
                "type": "MustacheTag",
              },
            ],
            "end": 92,
            "name": "p",
            "start": 76,
            "type": "Element",
          },
          {
            "data": "
      ",
            "end": 99,
            "raw": "
      ",
            "start": 92,
            "type": "Text",
          },
        ],
        "end": 105,
        "name": "div",
        "start": 0,
        "type": "Element",
      },
    ],
    "components": [],
    "directives": [],
    "helpers": Set {},
    "source": "<div>
        <p>{ first }</p>
        { second }
        { third }
        <p>{ forth }</p>
      </div>",
    "temps": 0,
    "type": 0,
  },
  "source": "<div>
        <p>{ first }</p>
        { second }
        { third }
        <p>{ forth }</p>
      </div>",
  "template": [],
  "type": 0,
}
`;
